src/ty_check/unify.rs

```diff
@@ -106,19 +190,9 @@ impl Ty {
         if Self::occurs_check(&IdentKinded::new(ident, Kind::Ty), self) {
             return Err(TyError::InfiniteType);
         }
-        if let Some(old) = constr_map
-            .ty_unifier
-            .insert(ident.name.clone(), self.clone())
-        {
-            if &old != self {
-                panic!(
-                    "Rebinding bound type variable.\n\
-                    Old: {:?}\n\
-                    New: {:?}",
-                    old, self
-                );
-            }
-        }
+        let mut self_ty = self.clone();
+        self_ty.substitute(&constr_map);
+        constr_map.ty_unifier.insert(ident.name.clone(), self_ty);
```

Comment:
Why is rebinding required now?

Reply:
A simple example is the following:
We try to unify `Add<$T, $T>` with `Add<$X, i32>` (or the other way round)
First we unify `$T` and `$X` => substitution [$X / $T] (or [$T / $X ])
then we unify `$T` and `i32` => result substitution should be in both cases: [i32 / $T, i32 / $X]

But because i wrote just now a test-case for this example, i noticed the it just works for one direction not for both. I will fix it in the next commit.
Maybe we dont really need rebinding and substitutions on bound values are enough. I will try it out. At the moment one test fails if i forbid rebinding bound type variables, but mabye it works without, if i fix the unification

Reply:
It seems like i didnt unterstand the details of the implementations, so i did implement the unifications for trait-mono-types ans structs wrong. This was the reason for my rebinding problems.
With a correct implemntation it works fine without rebindings. Also the implementation for tuples was wrong and for functions and references not totally correct.